src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCallQueue.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    // A mechanism for queuing JS-to-.NET calls so they aren't nested on the call stack and hence
+    // have the same ordering behaviors as in Blazor Server. This eliminates serveral inconsistency
+    // problems and bugs that otherwise require special-case solutions in other parts of the code.
+    //
+    // The reason for not using an actual SynchronizationContext for this is that, historically,
+    // Blazor WebAssembly has not enforced any rule around having to dispatch to a sync context.
```

Comment:
While this is factually true (as in it will not break in webassembly) this has not been the guidance that we've been giving customers (we always told them to use use `InvokeAsync` when "dispatching" to Blazor) and we shouldn't think of this as a scenario with any level of support "It's a bug" that we didn't catch. 

We don't want to encourage this practice since it breaks "universal" libraries and is not a guarantee that we can make in the future if we introduce multithreading when available. If a library chose to avoid the guidance despite our efforts I don't think we should be jumping hoops to avoid breaking them.

Reply:
This would be a really serious breaking change that we should only contemplate if there was no alternative, and after making some effort to quantify exactly how many apps would get hurt. Fortunately we have a straightforward way of not needing such a breaking change.

Reply:
> This would be a really serious breaking change that we should only contemplate if there was no alternative

This is detrimental for the Blazor ecosystem as a whole since it creates libraries that only work on Webassembly. It just "happens to work" on WASM because its currently single threaded, however we've made no guarantee that this would work. It's like relying on undefined behavior in a C compiler, the guidance is clear and if you choose not to follow it, your app might break when you update.

I think its best to address this sooner rather than later, the more we wait the bigger a problem this is going to be for people in the future. I don't think waiting until we are forced to do it is the right call here, since it will create a potentially much bigger problem in the future. And not just for Blazor users, but for the Blazor component ecosystem as a whole since the list of incompatible component libraries is likely to grow.

Reply:
For anyone else reading and future clarity, we discussed this and came to the conclusion that, until WebAssembly has true multithreading, it's not really meaningful to talk about enforcement of a sync context, since nested synchronous calls would already be on it anyway (and nested calls are precisely the thing we need to defer to fix the bugs this addresses).